File: 2019-03-03-169,229/watercow/169.py
# 使用count记录次数
# 遇到相同的数+1，遇到不同的-1
# 使用cur_num记录当前数，count为0时更换cur_num
# 最后的cur_num就是最多的num

class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return None

        count = 1
        cur_num = nums[0]

        for i in range(1, len(nums)):
            if nums[i] == cur_num:
                count += 1
            if nums[i] != cur_num:
                count -= 1
            if count == 0:
                cur_num = nums[i]
                count = 1

        return cur_num

File: 2019-03-03-169,229/lb/169.java
public class MajorityElement {
    /**
     * 因为题目已经指出，众数的数量大于n/2，笑到最后的只能是众数
     * @param nums
     * @return
     */
    public int majorityElement(int[] nums) {
        int count=0;
        int ret = 0;
        
        for(int num:nums){
            if (count == 0)
                ret = num;
            if (ret != num)
                count --;
            else
                count ++;
        }
        return ret;
    }
}
